doc/source/custom_directives.py

```diff
@@ -1252,6 +1255,45 @@ def pregenerate_example_rsts(
             )
 
 
+ray_prefix = "ray-"
+min_version = "1.11.0"
+
+
+def generate_version_url(version):
+    return f"https://docs.ray.io/en/{version}/"
+
+
+#
+# Gets the releases from git, sorts them in semver order,
+# and generates the JSON needed for the version switcher
+#
+def generate_versions_json():
+    try:
+        versions = []
+        tags = subprocess.check_output(["git", "tag"]).decode("utf-8").split()
```

Comment:
Is there a place to put this such that it only happens when building for real and not locally? I didn't have much luck finding a testing an alternative

Reply:
No, this is the place to do it. I'm not sure about the best way to do this, but here are some thoughts:

1. People who fork the ray repo often alias the upstream repo to `upstream`. For me, I can fetch all tags from the upstream repo with `git fetch upstream --tags`
2. If this is being run on a github runner, I believe there's a way to fetch tags in the options for `actions/checkout` since I think that does a shallow clone by default. Probably safest to do something in the github layer
3. I haven't searched for a way to get the upstream repo, but maybe someone else has posted on SO looking for the definitive way to find the root URL for a repository? That way you can find out whether someone has cloned a fork of the repo or whether they've cloned `ray-project/ray`.
4. You might be able to check for the presence of certain branches...but that seems like it might have edge cases that don't work?